lucene/core/src/java/org/apache/lucene/search/LRUQueryCache.java

```diff
@@ -867,12 +881,25 @@ public BulkScorer bulkScorer(LeafReaderContext context) throws IOException {
 
       if (docIdSet == null) {
         if (policy.shouldCache(in.getQuery())) {
+          boolean performSynchronousCaching = !(executor != null);
           // If asynchronous caching is requested, perform the same and return
           // the uncached iterator
-          if (executor != null) {
-            cacheAsynchronously(context, cacheHelper);
-            return in.bulkScorer(context);
-          } else {
+          if (!performSynchronousCaching) {
+            try {
+              cacheAsynchronously(context, cacheHelper);
+            } catch (RejectedExecutionException e) {
```

Comment:
Do we need to handle these exceptions? Isn't this equivalent to the caller's having specified a "run in caller" RejectedExecutionHandler? The caller may have specified a different policy, but here we are not respecting that

Reply:
If the caller has specified a different policy -- we should not be even getting here. `IndexSearcher#search` API does not advertise the behaviour for the case if the underlying `Executor` does not have enough capacity, nor does it consider the size of the executor during allocation of leaves. Both of the above facts lead to an implicit API contract where we are expected to gracefully handle the case of rejected execution. In case the user wishes to handle the exception themselves, they are free to override the default methods -- similar to what we advice for other methods in `IndexSearcher`.